usb/core.py

```diff
@@ -270,6 +270,9 @@ def __init__(self, strerror, error_code = None, errno = None):
         IOError.__init__(self, errno, strerror)
         self.backend_error_code = error_code
 
+class USBTimeoutError(USBError):
+    pass
```

Comment:
I would argue that for comparable behaviour we should return a `NotImplementedError` for openusb (and libusb0), too. Theoretically (in the wild it may not be so), users should already be catching NotImplementedErrors if they are not using specific backend deliberately. But either way it would introduce breaking changes that may need another version number if users want to keep the old behaviour. I will drop it but we could open another Issue for this (probably marked as enhancement).

Ok.

I'm now not sure what to do with libusb0? The error check method is more complicated, as it also checks for error messages and on special (?) cases falls back to Python `os` error codes... Should a check happen before? And which constant would I define? (Or use the Python `os` error code + string representation?)

Reply:
I agree in principle, which is why I still think that the OpenUSB should get the _non_ breaking improvements.  But I really don't want to push breaking changes to a deprecated target... other than actually removing it (i.e. PyUSB 2.0 will already remove OpenUSB).

Reply:
Regarding libusb0 _in general_, doesn't it just use standard/POSIX errors (i.e. `os.errno`), with the only special case being `ETRANSFER_TIMEDOUT` (which I mentioned above why I'm against checking for it)?

However, I missed the `NotImplementedError` case on libusb0: unlike OpenUSB, libusb0 should get this change _eventually..._  just not now, as it would also be a breaking change.  And I'm planning to have the next release be a minor one (new features, but no breaking changes).

Reply:
I thought that libusb0 may use the POSIX error codes but it also had a clause to check for (custom) error messages via the native library so I was not really sure ...

Reply:
Yeah, as far I could check, libusb0 indeed does use POSIX error codes... only libusb-win32 (libusb0 port to Windows) seems to have that one special case, but I couldn't really tell you any other errors aren't different as well (since they are coming, ultimately, from the Windows kernel).

That's why I think that for libusb0 we should only handle ETIMEDOUT (POSIX) as `USBTimeoutError`, and leave all other error codes as the current/default `USBError`.

Reply:
So, I got to code again ;-)

My solution for `libusb0` would be:
```diff
--- a/usb/backend/libusb0.py
+++ b/usb/backend/libusb0.py
@@ -14,11 +14,12 @@
 # limitations under the License.

 from ctypes import *
+import errno
 import os
 import usb.backend
 import usb.util
 import sys
-from usb.core import USBError
+from usb.core import USBError, USBTimeoutError
 from usb._debug import methodtrace
 import usb._interop as _interop
 import logging
@@ -426,6 +427,9 @@ def _check(ret):
                 errmsg = os.strerror(-ret)
         else:
             return ret
+
+    if ret is not None and -ret == errno.ETIMEDOUT:
+        raise USBTimeoutError(errmsg, ret, -ret)
     raise USBError(errmsg, ret)

 def _has_iso_transfer():
```

I noticed that the USBError does not get any errno parameter. Should I do this with the Timeout error, too?

In `libusb1` the errno mapping (here with `-ret`) was done with `LIBUSB_ERROR_TIMEOUT:errno.__dict__.get('ETIMEDOUT', None)`. That would catch the case that `errno.ETIMEDOUT` is not defined in the Python module. Can that happen in any way? If yes, the check above has to be a little more complex ... (Something like `... and "ETIMEDOUT" in errno.errorcode.values() and ...`)

If ok, then I will push the change to the pull request.